### DISCUSSION TITLE
Warning: "The previous version x.x.x.x contains obsolete procedure XXX that is still present in the current version y.y.y.y."

### DISCUSSION THREAD

**Discussion by rvanbekkum** [StefanMaron/BusinessCentral.LinterCop#140] [New Rule]
Add a new code analyzer rule that raises a warning if the previous version contains an obsoleted object/procedure/... that is still present in the current version.
The warning would be something like:
> The previous version x.x.x.x contains obsolete procedure XXX that is still present in the current version y.y.y.y.

**Comment by jwikman** [comment]
Isn't the common practice of obsoleted objects/procedures/... that you should keep them for a couple of majors? Since the customers of AppSource apps won't get updated automatically if we release a minor update, most probably won't notice the obsoletion until next major version. And then they might need some time to adapt.
I would say that removing them in the version right after they were set as `ObsoleteState = Pending` would be an exception, at least in my company... 

But I would love a rule that checked the `ObsoleteTag`, and if the major version of the `ObsoleteTag` is two majors ago or older it would throw out a warning. Maybe the number of majors should be configurable, I realized that Microsoft would set 3 here ([ref](https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-deprecation-timeline#how-long-time-from-obsoleted-to-deprecated))...

Eg. I've added `18.4` as `ObsoleteTag` (released in v18.4 of our AppSource app), when working on any release on v18 or v19 I do not want any warning, but when I change the major version of my app to `20.0` I would get the warning.

All this obviously assumes that you are using `[major].[minor]` as ObsoleteTag.

Thoughts?

**Reply by rvanbekkum** [reply]
Yes, I agree, you want to give other extensions a chance to adapt to the obsoletion. My initial intention with this rule was therefore also to at least get an indication that we have some running obsoletions in the latest `master` releases.

I think your additional suggestion of checking the `ObsoleteTag` attribute for major (and maybe even minor), under the assumption that you put version numbers in there like you said, would make the rule more usable, i.e., the warnings more meaningful.
